File: destack-ts/src/cli/console.ts
// ANSI color codes
const colors = {
  reset: "\x1b[0m",
  bold: "\x1b[1m",
  dim: "\x1b[2m",
  
  // foreground colors
  black: "\x1b[30m",
  red: "\x1b[31m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  blue: "\x1b[34m",
  magenta: "\x1b[35m",
  cyan: "\x1b[36m",
  white: "\x1b[37m",
  gray: "\x1b[90m",
  
  // bright foreground colors
  brightRed: "\x1b[91m",
  brightGreen: "\x1b[92m",
  brightYellow: "\x1b[93m",
  brightBlue: "\x1b[94m",
  brightMagenta: "\x1b[95m",
  brightCyan: "\x1b[96m",
  brightWhite: "\x1b[97m",
} as const;

type Color = keyof typeof colors;

/** Apply color to text */
export function color(text: string, ...styles: Color[]): string {
  const codes = styles.map(s => colors[s]).join("");
  return `${codes}${text}${colors.reset}`;
}

/** Print colored text to stdout */
export function print(text: string, ...styles: Color[]): void {
  console.log(color(text, ...styles));
}

/** Print text without newline */
export function write(text: string, ...styles: Color[]): void {
  process.stdout.write(color(text, ...styles));
}

/** Print error to stderr */
export function error(text: string): void {
  console.error(color(text, "red"));
}

/** Print warning */
export function warn(text: string): void {
  console.warn(color(text, "yellow"));
}

/** Print success message */
export function success(text: string): void {
  console.log(color(text, "green"));
}

/** Print info message */
export function info(text: string): void {
  console.log(color(text, "cyan"));
}

/** Clear the terminal */
export function clear(): void {
  process.stdout.write("\x1b[2J\x1b[0f");
}

/** Move cursor up n lines */
export function moveUp(lines: number = 1): void {
  process.stdout.write(`\x1b[${lines}A`);
}

/** Clear the current line */
export function clearLine(): void {
  process.stdout.write("\x1b[2K\r");
}

/** Create a simple box around content */
export function box(content: string, padding: number = 1): string {
  const lines = content.split("\n");
  const maxLength = Math.max(...lines.map(l => l.length));
  const width = maxLength + padding * 2;
  
  const result: string[] = [];
  
  // top border
  result.push("┌" + "─".repeat(width) + "┐");
  
  // content with padding
  for (const line of lines) {
    const paddedLine = line.padEnd(maxLength);
    result.push("│" + " ".repeat(padding) + paddedLine + " ".repeat(padding) + "│");
  }
  
  // bottom border
  result.push("└" + "─".repeat(width) + "┘");
  
  return result.join("\n");
}

/** Create a section header */
export function header(text: string, char: string = "="): string {
  const line = char.repeat(text.length);
  return `${line}\n${text}\n${line}`;
}

File: destack-rs/destack_cli/src/console/console.rs
//! Colorized console helpers.

use std::io::{self, Write};

/// Apply ANSI color/style to a string.
pub fn color(text: &str, code: &str) -> String {
    format!("\x1b[{code}m{text}\x1b[0m")
}

/// Print a line to stdout.
pub fn print(text: &str) {
    println!("{text}");
}

/// Print an info line to stdout in cyan.
pub fn info(text: &str) {
    println!("{}", color(text, "36"));
}

/// Print a success line to stdout in green.
pub fn success(text: &str) {
    println!("{}", color(text, "32"));
}

/// Print a warning line to stdout in yellow.
pub fn warn(text: &str) {
    println!("{}", color(text, "33"));
}

/// Print an error line to stderr in red.
pub fn error(text: &str) {
    eprintln!("{}", color(text, "31"));
}

/// Print text without newline to stdout.
pub fn write(text: &str) {
    let _ = io::stdout().write_all(text.as_bytes());
    let _ = io::stdout().flush();
}

/// Create a header rendered with an underline.
pub fn header(text: &str, underline: char) -> String {
    let line = underline.to_string().repeat(text.len());
    format!("{line}\n{text}\n{line}")
}

/// Render a simple ASCII/Unicode frame around content with optional title.
pub fn frame(content: &str, title: Option<&str>, padding: u8) -> String {
    // split content into lines
    let mut lines: Vec<&str> = if content.is_empty() {
        vec![""]
    } else {
        content.split('\n').collect()
    };

    // drop trailing empty line if content ended with a newline
    if content.ends_with('\n') && lines.last().is_some_and(|l| l.is_empty()) {
        lines.pop();
    }

    // count max width of content lines plus padding
    let content_width = lines.iter().map(|l| l.chars().count()).max().unwrap_or(0);
    let inner_width = content_width + (padding as usize * 2);
    let mut out = String::new();

    // top border sized to inner width
    if let Some(t) = title.filter(|t| !t.is_empty()) {
        let prefix = format!("─ {t} ");
        let used = prefix.chars().count();
        let fill = inner_width.saturating_sub(used);
        out.push('┌');
        out.push_str(&prefix);
        out.push_str(&"─".repeat(fill));
        out.push('┐');
    } else {
        out.push_str(&format!("┌{}┐", "─".repeat(inner_width)));
    }
    out.push('\n');

    // content lines with padding
    let pad_str = " ".repeat(padding as usize);
    for l in lines {
        let content_pad = inner_width.saturating_sub(l.chars().count() + (padding as usize * 2));
        out.push('│');
        out.push_str(&pad_str);
        out.push_str(l);
        out.push_str(&" ".repeat(content_pad));
        out.push_str(&pad_str);
        out.push('│');
        out.push('\n');
    }

    // bottom border
    out.push_str(&format!("└{}┘", "─".repeat(inner_width)));
    out
}
